[cutlass backend] cache filtered ops based on layouts

henrylhtsang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #153006
* __->__ #152580

Differential Revision: [D73972687](https://our.internmc.facebook.com/intern/diff/D73972687/)

Add cache to store the list of filtered ops for a specific shape + layout + dtype (aka hash on input_nodes).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

henrylhtsang committed:
[cutlass backend] cache filtered ops based on layouts



Differential Revision: [D73972687](https://our.internmc.facebook.com/intern/diff/D73972687/)

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] cache filtered ops based on layouts"


Differential Revision: [D73972687](https://our.internmc.facebook.com/intern/diff/D73972687/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -423,6 +430,8 @@ def __init__(
             [node.get_layout() for node in input_nodes]
         )
 
+        self.cache_key: str = create_inputs_key(self.input_nodes) + repr(layout)
```
@eellison this part confuses me. For autotuning results, we would cache by `create_inputs_key(self.input_nodes)` without using `layout` at all, which means layout is assumed to be always fixed?

eellison replied:
We fix input layouts, that's correct.

henrylhtsang replied:
> We fix input layouts, that's correct. Layouts should give fixed on lowering to mm.

ok I will revise to not use ` + repr(layout)`

henrylhtsang committed:
Update on "[cutlass backend] cache filtered ops based on layouts"


Differential Revision: [D73972687](https://our.internmc.facebook.com/intern/diff/D73972687/)

Add cache to store the list of filtered ops for a specific shape + layout + dtype (aka hash on input_nodes).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] cache filtered ops based on layouts"


Differential Revision: [D73972687](https://our.internmc.facebook.com/intern/diff/D73972687/)

Add cache to store the list of filtered ops for a specific shape + layout + dtype (aka hash on input_nodes).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
